_posts/2024-10-24-why-new-rtos.markdown

```diff
@@ -0,0 +1,212 @@
+When this call returns, the TLS compartment knows that this buffer is safe from time-of-check-to-time-of-use problems because the TCP/IP compartment cannot have captured it because the pointer does not have global permission.
+It knows that no data outside of that buffer can have been written, because the pointer is bounded.
+It knows that data left over in that buffer cannot have been read (because the pointer does not have read permission) and so it is safe from information disclosure.
+All of this is visible in the source code and is enforced with CHERI.
+
+It might be possible to get the same guarantees from an MPU by invoking a kernel to set up an MPU region with store-only permission covering an object owned by the TLS stack, but this would look very different.
+Trying to provide the same programmer model for both CHERI and MPU-based systems (or even those without an MPU) would require serious compromises.
+
+MMUs were not designed to provide abstractions for normal programmers.
+They were created to provide operating-system abstractions such as swapping and the ability to isolate processes or virtual machines from each other.
+MPUs were designed to allow an RTOS kernel to protect itself from untrusted tasks.
+They were both designed as tools for use by operating systems.
+
+CHERI, in contrast, was designed to provide a protection model that could be directly exposed into C and higher-level programming languages.
+It is designed to protect things that are visible in a language abstract machine as objects, not memory ranges.
+It is designed to protect them by putting the permissions (and bounds and integrity guarantees) in things that are exposed to programming languages as pointers, not as entries in an indirection table.
+
+## CHERI enables new abstractions
+
+There are a lot of abstractions in CHERIoT RTOS that are possible only because of CHERI.
+Most notably, we make a lot of use of the *sealing* mechanism from CHERI.
+Sealing is an operation that takes a CHERI capability representing something that a programmer would think of as a pointer and another that conveys the authority to seal an object as a specific type and produces a sealed capability.
+Once you have created a sealed capability, the only operation that you can do with it is unseal it with a permit-unseal capability that matches the type used for the sealing operation.
+
+If you didn't understand that, don't worry, the guarantees are more important than the mechanism.
+From a C programmer's perspective, this means that you can create opaque pointers that are tamper-proof and type safe.
+You can return a sealed pointer to an object from an API that you expose at a compartment boundary and use that as a handle.
+The caller can store this just as they would any other opaque pointer but cannot dereference it.
+When they pass it back (possibly indirectly), the compartment that created it can unseal it and be certain that it points to a value of the correct type.
```

Comment:
```suggestion
When they pass it back (possibly indirectly, through other compartments), your compartment that created it can unseal it and be certain that it points to a value of the correct type.
```